File: card-api/src/main/java/com/jean202/cardmizer/api/normalization/NormalizedTransaction.java
package com.jean202.cardmizer.api.normalization;

import java.util.Objects;
import java.util.Set;

public record NormalizedTransaction(String merchantCategory, Set<String> paymentTags) {
    public NormalizedTransaction {
        if (merchantCategory == null || merchantCategory.isBlank()) {
            throw new IllegalArgumentException("merchantCategory must not be blank");
        }
        paymentTags = Set.copyOf(Objects.requireNonNull(paymentTags, "paymentTags must not be null"));
    }
}

File: card-api/src/main/kotlin/com/jean202/cardmizer/api/normalization/NormalizedTransaction.kt
package com.jean202.cardmizer.api.normalization

class NormalizedTransaction(
    val merchantCategory: String,
    paymentTags: Set<String>,
) {
    val paymentTags: Set<String>

    init {
        require(merchantCategory.isNotBlank()) { "merchantCategory must not be blank" }
        this.paymentTags = paymentTags.toSet()
    }
}
